ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -765,6 +805,68 @@ module Winbind = struct
       hostname
 end
 
+module ClosestKdc = struct
+  let invalid_kdc = "invalid_kdc"
+  let periodic_update_task_name = "Update closest kdc"
+  let mtime_this ~name ~f =
+    try
+      let start = Mtime_clock.counter() in
+      f();
+      (name, Mtime_clock.count start)
+    with _ ->
+      debug "Failed to time %s" name;
+      (name, Mtime.Span.max_span)
+
+  let update_db ~domain ~kdc =
+  (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> fun value ->
+     (domain, kdc) :: List.remove_assoc domain value |> fun value ->
+     Db.Host.set_external_auth_configuration ~__context ~self ~value)
+  |> Server_helpers.exec_with_new_task "update domain closest kdc"
+
+  let from_db domain =
+    (fun __context ->
+     let self = Helpers.get_localhost ~__context in
+     Db.Host.get_external_auth_configuration ~__context ~self|> List.assoc_opt domain)
+  |> Server_helpers.exec_with_new_task "query domain closest kdc"
+
+  let lookup domain =
+    try
+      let krbtgt_sid = Wbinfo.sid_of_name (Printf.sprintf "%s@%s" krbtgt domain) |> maybe_raise in
+      let domain_netbios_name = Wbinfo.domain_name_of ~target_name_type:NetbiosName ~from_name: domain |> maybe_raise in
+
+      kdcs_of_domain domain
+      |> List.map (fun kdc -> debug "Got domain '%s' kdc '%s'" domain kdc; kdc)
+      |> List.map (fun kdc -> mtime_this ~name:kdc ~f:(fun() -> Ldap.query_user krbtgt_sid domain_netbios_name kdc |> maybe_raise))
+      |> List.filter (fun (_, span) -> Mtime.Span.equal Mtime.Span.max_span span|> not)
+      |> List.sort (fun (_, s1) (_, s2) -> Mtime.Span.compare s1 s2)
+      |> hd (Printf.sprintf "domain %s does not has valid kdc" domain)
+      |> fst
+    with _ ->
+      debug "Failed to lookup domain %s closest kdc" domain;
+      invalid_kdc
+
+  let update() =
+    try
+      Wbinfo.all_domain_netbis() |> maybe_raise
+      |> List.map (fun domain_netbios -> Wbinfo.domain_name_of ~target_name_type:Name ~from_name:domain_netbios |> maybe_raise)
+      |> List.map (fun domain -> domain, lookup domain)
+      |> List.filter (fun (_, kdc) -> kdc <> invalid_kdc)
+      |> List.iter (fun (domain, kdc) -> update_db ~domain ~kdc)
+    with e -> error "Failed to update domain kdc %s" (Printexc.to_string e)
```

Comment:
I would use the result monad here, rather than trying to converting any errors to exceptions

e.g.
```
let* domains = Wbinfo.all_domain_netbis () in ...
```

also if we fail to get the domain name of any 'domain_netbios', then we don't update the db at all; you may want to check that this is the right behaviour

Reply:
Yes, this is expected, Find the closest KDC is only best effort.
If closest KDC is not stored in xaipdb, xapi will just choose the first one as before.

Reply:
I have updated the code to use monad as possible,
But I would ike to leave this one, as I expect this function to return void other than Result, 
which is expected by the `Xapi_periodic_scheduler.add_to_queue`

Reply:
you can still use the `let*` bindings, even if you function returns unit, you just have to put it into a block, e.g:

```
let res : ('a, 'b) result =
  let* x = y in
  ...
  Ok x
in
maybe_raise res
```